pkg/resource/deploy/providers/registry.go

```diff
@@ -628,10 +628,24 @@ func (r *Registry) Diff(ctx context.Context, req plugin.DiffRequest) (plugin.Dif
 
 	// Diff the properties.
 	filteredNewInputs := FilterProviderConfig(req.NewInputs)
+	// Fixup the old inputs and outputs to contain "version", old versions of the CLI wouldn't have force saved it to
+	// state even though we know guarantee it's always in new inputs. This isn't _technically_ correct, we're just
+	// assuming the old state is the same version as the current state it might genuinely be different but there's no
+	// way for us to tell. But not restoring version can cause diffs even for the same version (see
+	// https://github.com/pulumi/pulumi-docker-build/issues/581).
```

Comment:
If the `version` is only missing from `OldOutputs`, we probably ought to use the value from the `OldInputs`. (Since the provider is not supposed to be able to change it.)